src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,248 @@
+package com.android.volley.cronet;
+
+import android.content.Context;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Header;
+import com.android.volley.Request;
+import com.android.volley.toolbox.AsyncHttpStack;
+import com.android.volley.toolbox.ByteArrayPool;
+import com.android.volley.toolbox.HttpResponse;
+import com.android.volley.toolbox.PoolingByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.chromium.net.CronetEngine;
+import org.chromium.net.CronetException;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataProviders;
+import org.chromium.net.UrlRequest;
+import org.chromium.net.UrlRequest.Callback;
+import org.chromium.net.UrlResponseInfo;
+
+public class CronetHttpStack extends AsyncHttpStack {
```

Comment:
In terms of what more you need to do here - I'd go line by line through HurlStack and HttpClientStack and make sure you cover any functionality that makes sense here. At a glance, I see:

- support for a UrlRewriter interface - this is only in HurlStack, but effectively possible in HttpClientStack as well by providing a custom HttpClient.
- handling request timeouts

Reply:
I added support for UrlRewriter interface. As we discussed, Cronet doesn't support passing in request timeouts, so we will hold off on that for now. 

I also noticed that the requests all have a Priority number, and we can set priority on the UrlRequest.Builder. Should we set the priority here, or are those two different things?

Reply:
Historically the priority has only been used to determine the order in which requests get pulled from Volley's internal queue and not propagated to the underlying HTTP library itself. That said, this is probably more because it wasn't really possible/meaningful before, since we blocked network threads on completing requests rather than sending a bunch of requests of different priorities to the queue and waiting for them to get back.

So I think it's actually probably reasonable to push the Request priority down to Cronet in addition to using it to prioritize requests in Volley's own queues. If need be, we could always provide a hook to clients to turn that off (or they could leave priorities at normal to make them all equivalent), but I'd wait until we hear a concrete use case for that.

Feel free to do this in a small follow up if it's non-trivial to convert between the priorities to prevent this PR from dragging on.